File: xkcd/src/main/java/xyz/jienan/xkcd/list/ListFilterDialogFragment.java
package xyz.jienan.xkcd.list;

import android.app.Dialog;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.appcompat.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckedTextView;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import xyz.jienan.xkcd.R;

public class ListFilterDialogFragment extends DialogFragment {

    private List<HashMap<String, String>> list;
    private int selection;
    private OnItemSelectListener itemSelectListener;
    private int[] filters;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        final int[] icons = {R.mipmap.ic_launcher_round, R.drawable.ic_heart_on, R.drawable.ic_thumb_on};
        list = new ArrayList<>();
        for (int i = 0; i < filters.length; i++) {
            HashMap<String, String> hashMap = new HashMap<>();
            hashMap.put("filter", getString(filters[i]));
            hashMap.put("iconRes", String.valueOf(icons[i]));
            list.add(hashMap);
        }
        FilterAdapter adapter = new FilterAdapter();

        builder.setAdapter(adapter, (dialog, which) -> {
            if (itemSelectListener != null) {
                itemSelectListener.onItemSelected(which);
            }
            dialog.dismiss();
        });

        return builder.create();
    }

    public void setItemSelectListener(OnItemSelectListener itemSelectListener) {
        this.itemSelectListener = itemSelectListener;
    }

    public void setSelection(int selection) {
        this.selection = selection;
    }

    public void setFilters(int[] filters) {
        this.filters = filters;
    }

    public interface OnItemSelectListener {
        void onItemSelected(int which);
    }

    private class FilterAdapter extends BaseAdapter {

        @Override
        public int getCount() {
            return list.size();
        }

        @Override
        public Object getItem(int position) {
            return list.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            View view = LayoutInflater.from(getContext()).inflate(R.layout.item_filter_dialog, parent, false);
            HashMap<String, String> map = list.get(position);
            ImageView ivFilter = view.findViewById(R.id.iv_filter);
            CheckedTextView tvFilter = view.findViewById(R.id.tv_filter);
            ivFilter.setImageResource(Integer.valueOf(map.get("iconRes")));
            tvFilter.setChecked(position == selection);
            tvFilter.setText(map.get("filter"));
            return view;
        }
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/list/ListFilterDialogFragment.kt
package xyz.jienan.xkcd.list

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.CheckedTextView
import android.widget.ImageView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import xyz.jienan.xkcd.R
import xyz.jienan.xkcd.list.activity.BaseListActivity
import java.util.*

class ListFilterDialogFragment() : DialogFragment() {

    companion object {
        private const val KEY_HAS_FAV = "hasFav"

        fun newInstance(hasFav: Boolean): ListFilterDialogFragment {
            val fragment = ListFilterDialogFragment()
            fragment.arguments = bundleOf(KEY_HAS_FAV to hasFav)
            return fragment
        }
    }

    private var list: MutableList<HashMap<String, Any>> = mutableListOf()

    var selection: Int = 0

    var itemSelectListener: OnItemSelectListener? = null

    var filters = intArrayOf()

    private val hasFav: Boolean
        get() = arguments?.getBoolean(KEY_HAS_FAV) != false

    private val icons: IntArray by lazy {
        if (hasFav) {
            intArrayOf(R.mipmap.ic_launcher_round, R.drawable.ic_heart_on, R.drawable.ic_thumb_on)
        } else {
            intArrayOf(R.mipmap.ic_launcher_round, R.drawable.ic_thumb_on)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(requireActivity())

        filters.filterIndexed { index, i -> hasFav || index != 1 }.forEachIndexed { index, res ->
            list.add(hashMapOf("filter" to getString(res), "iconRes" to icons[index]))
        }

        val adapter = FilterAdapter()

        builder.setAdapter(adapter) { dialog, which ->
            val selection = if (hasFav) {
                when (which) {
                    0 -> BaseListActivity.Selection.ALL
                    1 -> BaseListActivity.Selection.MY_FAVORITE
                    else -> BaseListActivity.Selection.PEOPLES_CHOICE
                }
            } else {
                when (which) {
                    0 -> BaseListActivity.Selection.ALL
                    else -> BaseListActivity.Selection.PEOPLES_CHOICE
                }
            }
            itemSelectListener?.onItemSelected(selection)
            dialog.dismiss()
        }

        return builder.create()
    }

    interface OnItemSelectListener {
        fun onItemSelected(selection: BaseListActivity.Selection)
    }

    private inner class FilterAdapter : BaseAdapter() {

        override fun getCount() = list.size

        override fun getItem(position: Int) = list[position]

        override fun getItemId(position: Int) = position.toLong()

        @SuppressLint("ViewHolder")
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view = LayoutInflater.from(context).inflate(R.layout.item_filter_dialog, parent, false)
            val map = list[position]
            val ivFilter = view.findViewById<ImageView>(R.id.iv_filter)
            val tvFilter = view.findViewById<CheckedTextView>(R.id.tv_filter)
            ivFilter.setImageResource(map["iconRes"]!! as Int)
            tvFilter.isChecked = position == selection
            tvFilter.text = map["filter"].toString()
            return view
        }
    }
}
